web/src/actions/admin.js

```diff
@@ -82,20 +56,48 @@ export const getRoles = () => dispatch => {
   return Promise.all([rolesPromise, activitiesPromise]);
 };
 
-export const ADMIN_GET_USERS_REQUEST = Symbol('admin : get users : request');
-export const ADMIN_GET_USERS_SUCCESS = Symbol('admin : get users : success');
-export const ADMIN_GET_USERS_ERROR = Symbol('admin : get users : error');
+export const ADMIN_CREATE_USER_REQUEST = Symbol(
+  'admin : create user : request'
+);
+export const ADMIN_CREATE_USER_SUCCESS = Symbol(
+  'admin : create user : success'
+);
+export const ADMIN_CREATE_USER_ERROR = Symbol('admin : create user : error');
 
-export const getUsers = () => dispatch => {
-  dispatch({ type: ADMIN_GET_USERS_REQUEST });
+export const createUser = (
+  { email, name, password, role, state },
+  { dispatchGetUsers = getUsers } = {}
+) => dispatch => {
+  dispatch({ type: ADMIN_CREATE_USER_REQUEST });
 
   return axios
-    .get('/users')
-    .then(res => {
-      dispatch({ type: ADMIN_GET_USERS_SUCCESS, data: res.data });
+    .post('/users', { email, name, password, role, state })
+    .then(() => {
+      dispatch({ type: ADMIN_CREATE_USER_SUCCESS });
+      dispatch(dispatchGetUsers());
     })
-    .catch(() => {
-      dispatch({ type: ADMIN_GET_USERS_ERROR });
+    .catch(e => {
+      let message = 'Unknown error creating user';
+      switch (e.response.data.error) {
+        case 'add-user-invalid':
+          message = 'Email and password are required';
+          break;
+        case 'add-user-email-exists':
+          message = 'A user with this email already exists';
+          break;
+        case 'add-user-weak-password':
+          message = 'The provided password is too weak';
+          break;
+        case 'add-user-invalid-phone':
+          message = 'The provided phone number is invalid';
+          break;
+        default:
+          break;
+      }
```

Comment:
I see what you mean about hooking UI directly to actions, but I also love that there's one place for all the error messaging. Cake, why can't we have you and eat you too?

Reply:
These actions now pass the error message along to the reducers and don't reject back to the UI.